layer/unittest/trace_event_log_tests.cc

```diff
@@ -27,8 +31,167 @@ using ::testing::IsEmpty;
 
 namespace performancelayers {
 namespace {
-// An `InstantEvent`. Adds a `scope` variable to the args indicating event's
-// visibility in the trace viewer.
+std::string ValueToJsonString(bool value) { return value ? "true" : "false"; }
+
+std::string ValueToJsonString(int64_t value) { return std::to_string(value); }
+
+std::string ValueToJsonString(const std::vector<int64_t> &values) {
+  std::stringstream json_str;
+  json_str << "[";
+  size_t e = values.size();
+  for (size_t i = 0; i != e; ++i) {
+    const char *delimiter = i < e - 1 ? "," : "";
+    json_str << "\"0x" << std::hex << values[i] << "\"" << delimiter;
+  }
+  json_str << "]";
+  return json_str.str();
+}
+
+// Converts the duration to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(Duration value) {
+  return std::to_string(value.ToMilliseconds());
+}
+
+// Converts the timestamp to milliseconds, the defualt time unit in the `Trace
+// Event` format.
+std::string ValueToJsonString(TimestampClock::time_point value) {
+  return std::to_string(ToUnixMillis(value));
+}
+
+// Appends all given attributes, including the type-specific attribute, to the
+// stream.
+void TraceArgsToJsonString(const std::vector<Attribute *> &args,
+                           std::ostringstream &json_str) {
+  json_str << ", " << std::quoted("args") << " : { ";
+  for (size_t i = 0, e = args.size(); i < e; ++i) {
```

Comment:
nit: prefer range-based for loops

Reply:
I forgot to put the comma separator at the end. I needed the normal loop for that.